### PR TITLE
Split ir nodes for variable/field access into load/store

### PR DESCRIPTION
**Pull request by jdconrad**
This change splits the ir nodes for variable/field accesses into two separate nodes, one for load and one for store. This removes the need for an assignment node as the store nodes replace it as the root. It also means all nodes simply have a write method instead of having separate methods for load and store to support the different types of bytecode required for this. This also means that compound assignment is no longer specialized on the assignment node which had to be overly coupled with the previous version of the variable/field accessor nodes.

Store nodes now have two children. The first child is the prefix chain. The second child is the value to store.  The prefix chain is the set of nodes that represent `a.b` in the accessor chain of `a.b.c`. The value node can be any value that is legally castable to type of the store node.

Compound assignments are now setup where the value node of a store node is a binary math node. The left side of the binary math node is a dup node with a prefix root node as a child. The right side of the binary math node is the value node.

### CODE REVIEW

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java [hunk 1]
```diff
@@ -386,6 +401,89 @@ protected ExpressionNode injectCast(AExpression userExpressionNode, ScriptScope
         return irCastNode;
     }
 
+    protected ExpressionNode buildLoadStore(int accessDepth, Location location, boolean isNullSafe,
+            ExpressionNode irPrefixNode, ExpressionNode irIndexNode, ExpressionNode irLoadNode, StoreNode irStoreNode) {
+
+        ExpressionNode irExpressionNode;
+
+        if (irStoreNode == null) {
+            ExpressionNode irAccessNode;
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:410]: // This is a load.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java [hunk 2]
```diff
@@ -386,6 +401,89 @@ protected ExpressionNode injectCast(AExpression userExpressionNode, ScriptScope
...
+                BinaryNode irParentNode = new BinaryNode();
+                irParentNode.setLocation(location);
+                irParentNode.setExpressionType(irLoadNode.getExpressionType());
+                irParentNode.setLeftNode(irPrefixNode);
+                irParentNode.setRightNode(irAccessNode);
+
+                irExpressionNode = irParentNode;
+            }
+        } else {
+            ExpressionNode irAccessNode;
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:444]: // This is a store

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java [hunk 3]
```diff
@@ -386,6 +401,89 @@ protected ExpressionNode injectCast(AExpression userExpressionNode, ScriptScope
...
+
+        ExpressionNode irExpressionNode;
+
+        if (irStoreNode == null) {
+            ExpressionNode irAccessNode;
+
+            if (irIndexNode == null) {
+                irAccessNode = irLoadNode;
+            } else {
+                BinaryNode irBinaryNode = new BinaryNode();
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:415]: // Array access

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java [hunk 4]
```diff
@@ -386,6 +401,89 @@ protected ExpressionNode injectCast(AExpression userExpressionNode, ScriptScope
...
+                nullSafeSubNode.setChildNode(irAccessNode);
+                nullSafeSubNode.setLocation(location);
+                nullSafeSubNode.setExpressionType(irAccessNode.getExpressionType());
+                irAccessNode = nullSafeSubNode;
+            }
+
+            if (irPrefixNode == null) {
+                irExpressionNode = irAccessNode;
+            } else {
+                BinaryNode irParentNode = new BinaryNode();
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:435]: // Variable node

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java [hunk 5]
```diff
@@ -386,6 +401,89 @@ protected ExpressionNode injectCast(AExpression userExpressionNode, ScriptScope
...
+            if (isNullSafe) {
+                NullSafeSubNode nullSafeSubNode = new NullSafeSubNode();
+                nullSafeSubNode.setChildNode(irAccessNode);
+                nullSafeSubNode.setLocation(location);
+                nullSafeSubNode.setExpressionType(irAccessNode.getExpressionType());
+                irAccessNode = nullSafeSubNode;
+            }
+
+            if (irPrefixNode == null) {
+                irExpressionNode = irAccessNode;
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:433]: // Variable node

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java (hunk 4 above)
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:435]: // Field and brace node

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java [hunk 6]
```diff
@@ -386,6 +401,89 @@ protected ExpressionNode injectCast(AExpression userExpressionNode, ScriptScope
...
+                irParentNode.setLeftNode(irPrefixNode);
+                irParentNode.setRightNode(irAccessNode);
+
+                irExpressionNode = irParentNode;
+            }
+        } else {
+            ExpressionNode irAccessNode;
+
+            if (irIndexNode == null) {
+                irAccessNode = irPrefixNode;
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:447]: // Standard field

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java [hunk 7]
```diff
@@ -386,6 +401,89 @@ protected ExpressionNode injectCast(AExpression userExpressionNode, ScriptScope
...
+            } else {
+                BinaryNode irBinaryNode = new BinaryNode();
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:449]: // Storing brace operation

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java [hunk 8]
```diff
@@ -386,6 +401,89 @@ protected ExpressionNode injectCast(AExpression userExpressionNode, ScriptScope
...
+                irBinaryNode.setLocation(location);
+                irBinaryNode.setExpressionType(void.class);
+                irBinaryNode.setLeftNode(irPrefixNode);
+                irBinaryNode.setRightNode(irIndexNode);
+
+                irAccessNode = irBinaryNode;
+            }
+
+            if (irLoadNode == null) {
+                if (irAccessNode != null) {
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:459]: // Basic store

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java [hunk 9]
```diff
@@ -386,6 +401,89 @@ protected ExpressionNode injectCast(AExpression userExpressionNode, ScriptScope
...
+                    ((StoreAccessNode)irStoreNode).setAccessNode(irAccessNode);
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:460]: // Not a compound assignment

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java [hunk 10]
```diff
@@ -386,6 +401,89 @@ protected ExpressionNode injectCast(AExpression userExpressionNode, ScriptScope
...
+                }
+            } else {
+                if (irAccessNode != null) {
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:463]: // Compound assignment

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java [hunk 11]
```diff
@@ -386,6 +401,89 @@ protected ExpressionNode injectCast(AExpression userExpressionNode, ScriptScope
...
+                    DupNode dupNode = new DupNode();
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:464]: // Compound assignment

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java [hunk 12]
```diff
@@ -710,27 +808,93 @@ public void visitBreak(SBreak userBreakNode, ScriptScope scriptScope) {
...
-        irAssignmentNode.setCat(scriptScope.getCondition(userAssignmentNode, Concatenate.class));
-        irAssignmentNode.setThere(upcast);
-        irAssignmentNode.setBack(downcast);
-
-        scriptScope.putDecoration(userAssignmentNode, new IRNodeDecoration(irAssignmentNode));
+
+        StoreNode irStoreNode;
+        ExpressionNode irValueNode = injectCast(userAssignmentNode.getRightNode(), scriptScope);
+
+        if (compoundType != null) {
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:818]: Add comments to help follow to the leaves.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java [hunk 13]
```diff
@@ -710,27 +808,93 @@ public void visitBreak(SBreak userBreakNode, ScriptScope scriptScope) {
...
-        irAssignmentNode.setOperation(userAssignmentNode.getOperation());
-        irAssignmentNode.setRead(scriptScope.getCondition(userAssignmentNode, Read.class));
-        irAssignmentNode.setCat(scriptScope.getCondition(userAssignmentNode, Concatenate.class));
-        irAssignmentNode.setThere(upcast);
-        irAssignmentNode.setBack(downcast);
-
-        scriptScope.putDecoration(userAssignmentNode, new IRNodeDecoration(irAssignmentNode));
+
+        StoreNode irStoreNode;
+        ExpressionNode irValueNode = injectCast(userAssignmentNode.getRightNode(), scriptScope);
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:816]: // For upcast, downcast (int i = 5; i += 2.0;)

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java [hunk 14]
```diff
@@ -710,27 +808,93 @@ public void visitBreak(SBreak userBreakNode, ScriptScope scriptScope) {
...
+            } else {
+                CastNode irCastNode = new CastNode();
+                irCastNode.setLocation(irBinaryMathNode.getLocation());
+                irCastNode.setExpressionType(downcast.targetType);
+                irCastNode.setCast(downcast);
+                irCastNode.setChildNode(irBinaryMathNode);
+                irStoreNode.setChildNode(irCastNode);
+            }
+
+            if (read) {
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:837]: Bigger comment that says what we have to do to support it.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java [hunk 15]
```diff
@@ -1235,16 +1403,40 @@ public void visitSymbol(ESymbol userSymbolNode, ScriptScope scriptScope) {
...
+            staticNode.setLocation(userSymbolNode.getLocation());
+            staticNode.setExpressionType(staticType);
+            irExpressionNode = staticNode;
         } else if (scriptScope.hasDecoration(userSymbolNode, ValueType.class)) {
-            VariableNode irVariableNode = new VariableNode();
-            irVariableNode.setLocation(userSymbolNode.getLocation());
-            irVariableNode.setExpressionType(scriptScope.getDecoration(userSymbolNode, ValueType.class).getValueType());
-            irVariableNode.setName(userSymbolNode.getSymbol());
-            irExpressionNode = irVariableNode;
+            boolean read = scriptScope.getCondition(userSymbolNode, Read.class);
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:1411]: Dot, symbol, brace must handle loads and/or stores.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java (hunk 1 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:410]: Done.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java (hunk 2 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:444]: Done.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java (hunk 5 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:433]: Done.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java (hunk 3 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:415]: Done.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java (hunk 4 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:435]: Done.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java (hunk 6 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:447]: Done.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java (hunk 7 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:449]: Done.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java (hunk 8 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:459]: Done.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java (hunk 9 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:460]: Done.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java (hunk 11 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:464]: Done.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java (hunk 12 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:818]: Done.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java (hunk 13 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:816]: Done.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java (hunk 14 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:837]: Done.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java (hunk 15 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:1411]: Done.

**Comment by jdconrad** [comment]
@stu-elastic Thank you for the initial review. I've added comments as appropriate based on your feedback.

**Review by stu-elastic — APPROVED** [review]
looking fine, Mr. Conradson.

**Comment by jdconrad** [comment]
@stu-elastic Thanks for the review!

**Pull request merged**